reference/ast-examples.md

```diff
@@ -0,0 +1,147 @@
+
+@AnAnnotation
+public class A {
+
+    List<Integer> a = new ArrayList<>();
+
+    int foo() {
+        int a = 1 + 2, b = 3, c;
+        this.a = this.a.stream()
+                .map(it -> it + 1)
+                .collect(Collectors.toList());
+        return a;
+    }
+}
+
+@interface AnAnnotation {}
+```
+
+### Binary
+
+A [Binary element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L597-L705) is any operation with a left and right side (usually separated by something like a `+`).
+
+![Binary Element Example](<../.gitbook/assets/Binary.png>)
+
+### Block 
+
+A [Block element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L712-L851) is the code contained between curly braces. Blocks can be nested inside of each other.
+
+![Block Element Example](<../.gitbook/assets/Block.png>)
+
+### ClassDeclaration
+
+A [ClassDeclaration element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1062-L1336) contains all of the code for any Java class. 
+
+![ClassDeclaration Element Example](<../.gitbook/assets/ClassDeclaration.png>)
+
+### CompilationUnit
+
+A [CompilationUnit element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1342-L1545) is the root of the Java AST. All other elements are contained inside of this. 
+
+![CompilationUnit Element Example](<../.gitbook/assets/CompilationUnit.png>)
+
+### Expression
+
+An [Expression element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/Expression.java) is anything that returns a value. `MethodInvocation`, `Identifier`, and `Binary` are all examples of expressions. Please note that some AST elements such as `MethodInvocation` are both a [Statement](#statement) and an Expression.
+
+In the below code, only some of the expressions are highlighted as expressions can often have many expressions inside of them and it would be too difficult to read if all of them were highlighted. For instance, `import java.util.ArrayList` is many expressions (`java`, `util`, `ArrayList`, `java.util`, and `java.util.ArrayList`).
+
+![Expression Element Example](<../.gitbook/assets/Expression.png>)
+
+### FieldAccess
+
+A [FieldAccess element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1827-L1953) is any fully qualified name. Often times, these are package or import statements, but they can also appear in code as something like: `this.foo`.
+
+![FieldAccess Element Example](<../.gitbook/assets/FieldAccess.png>)
+
+### Identifier
+
+An [Identifier element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L2310-L2343) is any name in the code (class names, variable names, method names, etc).
```

Comment:
Good place to mention that `J.Identifier.getFieldType()` can be used to tell what class the identifier is a field on. `null` for identifiers which are not fields.